base/steer/FairRun.cxx

```diff
@@ -79,14 +79,29 @@ void FairRun::AddTask(FairTask* t)
 {
   fTask->Add(t);
   fNTasks++;
+  StoreTaskNames(t);
+}
+//_____________________________________________________________________________
+void FairRun::StoreTaskNames(const FairTask* t)
+{
+  if(NULL==t) return;
   fFileHeader->AddTaskClassName(t->ClassName());
+  TList* l=t->GetListOfTasks();
+  if (NULL==l) return;
+  for(TObject* t2: *l )
+  {
+    StoreTaskNames(static_cast<FairTask*>(t2));
+  }
+  return;
 }
 //_____________________________________________________________________________
 void FairRun::SetTask(FairTask* t)
 {
   delete fTask;
   fTask = t;
-  fFileHeader->AddTaskClassName(t->ClassName());
+  TList* l = t->GetListOfTasks();
+  if(NULL != l) l->Clear();
```

Comment:
I think we have also to reset the counter here (fNTasks=0) when the list is cleared.